sql: TestQueryCache failed

[(sql).TestQueryCache failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2107711&tab=buildLog) on [master@e9a4f83e3eee59510f97db2c6e0df9b57cf6b944](https://github.com/cockroachdb/cockroach/commits/e9a4f83e3eee59510f97db2c6e0df9b57cf6b944):

```
=== RUN   TestQueryCache
    TestQueryCache: test_log_scope.go:77: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryCache740951531
    TestQueryCache: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestQueryCache (2.93s)
=== RUN   TestQueryCache/group/statschange
=== PAUSE TestQueryCache/group/statschange
=== CONT  TestQueryCache/group/statschange
        --- FAIL: TestQueryCache/group/statschange (1.18s)
=== RUN   TestQueryCache/group
    --- FAIL: TestQueryCache/group (0.00s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestQueryCache PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQueryCache.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestQueryCache failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2108376&tab=buildLog) on [master@e9a4f83e3eee59510f97db2c6e0df9b57cf6b944](https://github.com/cockroachdb/cockroach/commits/e9a4f83e3eee59510f97db2c6e0df9b57cf6b944):

```
=== RUN   TestQueryCache
    TestQueryCache: test_log_scope.go:77: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryCache343816065
    TestQueryCache: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestQueryCache (1.16s)
=== RUN   TestQueryCache/group/statschange
=== PAUSE TestQueryCache/group/statschange
=== CONT  TestQueryCache/group/statschange
        --- FAIL: TestQueryCache/group/statschange (0.31s)
=== RUN   TestQueryCache/group
    --- FAIL: TestQueryCache/group (0.00s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestQueryCache PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQueryCache.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestQueryCache failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2111695&tab=buildLog) on [master@b8a50cc4d062293915969cdc83e3ec4d057cede5](https://github.com/cockroachdb/cockroach/commits/b8a50cc4d062293915969cdc83e3ec4d057cede5):

```
=== RUN   TestQueryCache
    TestQueryCache: test_log_scope.go:77: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryCache485450181
    TestQueryCache: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestQueryCache (1.89s)
=== RUN   TestQueryCache/group/statschange
=== PAUSE TestQueryCache/group/statschange
=== CONT  TestQueryCache/group/statschange
        --- FAIL: TestQueryCache/group/statschange (0.60s)
=== RUN   TestQueryCache/group
    --- FAIL: TestQueryCache/group (0.00s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestQueryCache PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQueryCache.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/plan_opt_test.go
+++ b/pkg/sql/plan_opt_test.go
@@ -90,6 +90,20 @@ func (h *queryCacheTestHelper) AssertStats(tb *testing.T, expHits, expMisses int
 	assert.Equal(tb, expMisses, misses, "misses")
 }
 
+// CheckStats is similar to AssertStats, but returns an error instead of
+// failing the test if the actual stats don't match the expected stats.
+func (h *queryCacheTestHelper) CheckStats(tb *testing.T, expHits, expMisses int) error {
+	tb.Helper()
+	hits, misses := h.GetStats()
+	if expHits != hits {
+		return errors.Errorf("expected %d hits but found %d", expHits, hits)
+	}
+	if expMisses != misses {
+		return errors.Errorf("expected %d misses but found %d", expMisses, misses)
+	}
+	return nil
+}
+
 func TestQueryCache(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -347,8 +361,17 @@ SELECT cte.x, cte.y FROM cte LEFT JOIN cte as cte2 on cte.y = cte2.x`, j)
 			h.AssertStats(t, 1 /* hits */, 1 /* misses */)
 			r0.Exec(t, "CREATE STATISTICS s FROM t")
 			h.AssertStats(t, 1 /* hits */, 1 /* misses */)
-			r1.CheckQueryResults(t, "SELECT * FROM t", [][]string{{"1", "1"}})
-			h.AssertStats(t, 1 /* hits */, 2 /* misses */)
+			hits := 1
+			testutils.SucceedsSoon(t, func() error {
+				// The stats cache is updated asynchronously, so we may get some hits
+				// before we get a miss.
+				r1.CheckQueryResults(t, "SELECT * FROM t", [][]string{{"1", "1"}})
+				if err := h.CheckStats(t, hits, 2 /* misses */); err != nil {
+					hits++
+					return err
+				}
+				return nil
+			})
 		})
 
 		// Test that a schema change triggers cache invalidation.

--- a/pkg/sql/stats/stats_cache_test.go
+++ b/pkg/sql/stats/stats_cache_test.go
@@ -377,7 +377,7 @@ func TestCacheWait(t *testing.T) {
 		before := sc.mu.numInternalQueries
 
 		id := tableIDs[rand.Intn(len(tableIDs))]
-		sc.RefreshTableStats(ctx, id)
+		sc.InvalidateTableStats(ctx, id)
 		// Run GetTableStats multiple times in parallel.
 		var wg sync.WaitGroup
 		for n := 0; n < 10; n++ {

